Deprecate Error's `to_display_text` builtin methods

hubertp opened the pull request:
### Pull Request Description

The primary motivation for this change was
https://github.com/enso-org/enso/issues/6248, which requested the possibility of defining `to_display_text` methods of common errors via regular method definitions. Until now one could only define them via builtins.

To be able to support that, polyglot invocation had to report `to_display_text` in the list of (invokable) members, which it didn't. Until now, it only considered fields of constructors and builtin methods. That is now fixed as indicated by the change in `Atom`.

Closes #6248.

### Important Notes

Once most of builtins have been translated to regular Enso code, it became apparent how the usage of `.` at the end of the message is not consistent and inflexible. The pure message should never follow with a dot or it makes it impossible to pretty print consistently for the purpose of error reporting. Otherwise we regularly end up with errors ending with `..` or worse. So I went medieval on the reasons for failures and removed all the dots.
The overall result is mostly the same except now we are much more consistent.

Finally, there was a bit of a good reason for using builtins as it simplified our testing.
Take for example `No_Such_Method.Error`. If we do not import `Errors.Common` module we only rely on builtin error types. The type obviously has the constructor but it **does not have** `to_display_text` in scope; the latter is no longer a builtin method but a regular method. This is not really a problem for users who will always import stdlib but our tests often don't. Hence the number of changes and sometimes lack of human-readable errors there.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- ~[ ] The documentation has been updated, if necessary.~
- ~[ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - ~[ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~

hubertp committed:
Deprecate Error's `to_display_text` builtin methods

The primary motivation for this change was
https://github.com/enso-org/enso/issues/6248, which requested the
possibility of defining `to_display_text` methods of common errors
via regular method definitions, not only via builtins.

To be able to support that, polyglot invocation had to report
`to_display_text` in the list of (invokable) members, which it didn't.
Until now, it only considered fields of constructors and builtin
methods. That is now fixed.

Once most of builtins have been translated to regular Enso code, it
became apparent how the usage of `.` at the end of the message is not
consistent and inflexible. The pure message should never follow with a
dot or it makes it impossible to pretty print consistently for the
purpose of error reporting. Otherwise we regularly end up with errors
ending in `..` or worse. So I went medieval on the reasons for failures
and removed all the dots.
The overall result is mostly the same except now we are much more
consistent.

Finally, there was a bit of a good reason for using builtins as it
simplified our testing.
Take for example `No_Such_Method.Error`. If we do not import `Errors.Common`
module we only rely on builtin error types. The type obviously has the
constructor but it **does not have** `to_display_text` in scope; the
latter is no longer a builtin method but a regular method.
This is not really a problem for users who will always import stdlib
but our tests often don't. Hence the number of changes and sometimes
lack of human-readable errors there.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Errors/Common.enso:
```diff
@@ -49,6 +54,18 @@ type Type_Error
+       Infer the type of the actual value in a human-readable format.
+    type_of_actual self =
+        case (Meta.type_of self.actual) of
+            Compile_Error.Error _ -> self.actual.to_display_text
```
Wait.

A `Compile_Error` is _returned_ as a _value_ and not a panic/dataflow error??? That is so weird. Why is that? This just seems wrong.

hubertp replied:
It's been like that for a while https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java#L181

radeusgd replied:
Yes, I see that. But that is totally strange.
Can we change it?

hubertp replied:
Sure.

JaroslavTulach replied:
I'd say that unknown types in Enso should have type `Any`. Because if you try to call a method on them, you can call any instance method of `Any` and the dispatch is going to work. @kustosz usually advises me when it comes to these `Any`/polyglot issues.

radeusgd commented on engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeErrorsTest.scala:
```diff
@@ -175,7 +175,7 @@ class RuntimeErrorsTest
         contextId,
         xId,
         Api.ExpressionUpdate.Payload.Panic(
-          "Compile error: The name `undefined` could not be found.",
+          "Compile_Error.Error",
           Seq(xId)
```
I'm slightly worried about losing these.

Are we sure that everything is wired in correctly that in production we will get the human readable errors? Getting `Compile_Error.Error` will be bad.

Cannot we import the part of stdlib responsible for the `to_display_text` here?

I think that ensuring that the errors are sensible in production is pretty important, so I'd advocate for just importing the full stdlib here. The compilation time is now much shorter, I think we can pay the extra 30s on the CI for ensuring that errors are correct.

radeusgd replied:
Especially as, we were quite often having bugs in our `to_display_text`. Nothing better than your code erroring when trying to display the error, so that you can error while you error and do not know the true error.

I think it's really good to have them tested.

hubertp replied:
See my comment in `Important Notes` section.

If something is not a builtin and you don't import the corresponding type from stdlib then how is it supposed to know about that method? All it knows is just builtins.
Let's be honest, IDE will always import `Standard.Base.Main` which should and does import those, so you won't see those in real life. Those test cases often don't import stdlib, for brevity, hence the error. That is also why I added sometimes a `pretty print` version, even if it is not really necessary.

A similar problem was with `Any.Any` during the equals work. That's why all our stdlib and tests are now _polluted_ with that import.

radeusgd replied:
I meant a bit different things, and I was wrong on one.

1. Indeed I'm not worried about IDE not importing the Base, but rather if the import is properly picked. I.e. if these are some internal methods, do they 'see' this import. But truth is - if we added the import here, we wouldn't test it properly either I guess. So I guess ignore this part.
2. Still, we tend to have had mistakes in the implementation of `to_display_text` which made it fail at runtime. We definitely do not want this to happen. So we need to test it _somewhere_. I thought this is an OK place for that. But if you think we can put those tests somewhere else, that's also OK. It is just, we had plans to add tests for `to_display_text` of _every_ error reported, to make sure there are no regressions in that regard (these regressions tended to happen _too_ often). So our goal is to add more tests for `to_display_text` of errors, and so that's why I'm opposed to doing essentially the reverse - removing some of the only tests that we currently have. If you need to remove them, can we have some tests in our Enso `Tests` which check that `to_display_text` of these errors does not fail at runtime?

hubertp replied:
re 2: but I didn't ignore testing the human-readable version of errors. See for example `RuntimeServerTest`. I'm fine to add a similar version here, in `RuntimeErrorsTes`, though the whole calculation of locations is rather cumbersome. That's why I skipped it in this case.

radeusgd replied:
Oh! Sorry, the diff was very large there and I didn't see this part. My bad. Looks all good then.

hubertp replied:
Added tests with human-readable text as well.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java:
```diff
@@ -132,17 +135,26 @@ public boolean hasMembers() {
+    if (members != null) {
+      allMembers.addAll(members.keySet());
+    }
+    members = constructor.getType().getDefinitionScope().getMethods().get(constructor.getType());
```
Membes are a union of what is in `constructor.getDefinitionScope()` and `constructor.getType().getDefinitionScope()` - ha: what's the difference? Why there is a difference?

hubertp replied:
Consider `Forbidden_Operation.Error`:
The former has `operation` field (and any builtins associated with the type). The latter (now) has `to_display_text` that is the method of `Forbidden_Operation`.

JaroslavTulach replied:
I don't see an explicit test for this behavior. Adding a test proving the above via `org.graalvm.polyglot.Value` might be beneficial.

hubertp replied:
@JaroslavTulach the test is that this transition from builtin method to regular enso methods is actually possible :)
Without this change, all those `to_display_text` methods would simply be ignored and we would be getting `XYZ_Error.Error` even if stdlib was imported. That is clearly not the case, given the number of tests I've added.

JaroslavTulach replied:
That is an indirect, integration test. I was seeking for a unit test that would check 

> The former has `xyz` field. The latter (now) has `abc` that is the method of the type - both are now members

Something clear & simple (not hundreds of cryptic failures across the codebase) that would help me understand what is the desired behavior.

hubertp replied:
done

JaroslavTulach reviewed (approved):
Removal of 15 builtins is good of its own. If the whole things holds together - go on.

hubertp committed:
type_of handles better some unexpected values

type_of "hole in an atom" is now `Nothing` with a warning.
type_of unresolved symbol will return a dataflow error, which allows us
to check it better in the text representation of `Type_Error`.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java:
```diff
@@ -116,6 +116,16 @@ Object doPolyglotNumber(
+  @Specialization
+  Object doholeInAtom(AtomWithAHoleNode.HoleInAtom holeInAtom) {
+    Builtins builtins = EnsoContext.get(this).getBuiltins();
+    return Warning.attach(
+        EnsoContext.get(this),
+        builtins.nothing(),
+        Text.create("Hole in an atom doesn't have a type before being filled"),
+        null);
```
That's OK, but wouldn't it be better to return a dataflow error?

Right now
```
Meta.type_of Nothing == Nothing == Meta.type_of atom.hole
```
so it is not easy to distinguish the situation where a hole in atom was queried vs where a `Nothing` type was. Of course one can do so by checking the warnings, but IMO a 'slightly more severe' dataflow error would suit better here. IMO it is an error to query a type of a _for all intents and purposes **uninitialized**_ value.

hubertp replied:
It's one of those things that would be nice if it was properly spec'ed. I will see if something breaks if I change it.

hubertp committed:
add more tests

hubertp committed:
remove special treatment for HoleInAtom in type of

hubertp committed:
add unit test for Atom's members

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/interpreter/test/TypeMembersTest.java:
```diff
@@ -87,6 +88,21 @@ public void checkAtomMembers() throws Exception {
+  public void ensureNonBuiltinMembersArePresent() throws Exception {
+    final URI uri = new URI("memory://how_long.enso");
+    final Source src = Source.newBuilder("enso", """
+    from Standard.Base.Errors.Common import Compile_Error
```
Nitpick: I was expecting the type to be defined here, not imported from StdLib. Something like:
```
type Cmon
  Error msg

  to_text self = "Cmon: "+self.msg
```

hubertp replied:
Not really, no. That actually works as expected without my change since it returns `["msg", "to_text"]`.
The crux is in the type being a builtin type.

hubertp replied:
You might also ask why. Well, it's because the `definitionScope` of the constructor is in builtins module, while in `Cmon.Error` it is in that `Cmon` one. Builtins are... special.

JaroslavTulach replied:
If I forced you to write a "unit" test for something that can't be unit tested, then I deeply apologize. But to not give up so easily. Would:

```
@Builtin_Type
type Compile_Error
    Error message
```
rather than `from Standard.Base.Errors.Common import Compile_Error` in the test work? But it'd be cheating and not a unit test anyway...

hubertp replied:
No worries. I will check that suggestion

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/interpreter/test/TypeMembersTest.java:
```diff
@@ -87,6 +88,21 @@ public void checkAtomMembers() throws Exception {
+
+    var module = ctx.eval(src);
+    var compileError = module.invokeMember("eval_expression", "v");
+    assertEquals("all members", compileError.getMemberKeys(), Set.of("to_display_text", "message"));
```
Shouldn't this also include members contributed by `Any` or other super types?

hubertp replied:
Wouldn't that then apply also to other places like `GetTypeMethodsNode`?

hubertp replied:
If we want to expand, I'd really like to consider doing it in yet another PR. This is already getting hard to review given the number of changes. I will work on followup.

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/6248-spurious-to-display-text-builtins
